### PR TITLE
Clarified the two main types of exceptions thrown in HttpClient

### PR DESCRIPTION
**Pull request by Toflar**
As learned from @nicolas-grekas in https://github.com/symfony/symfony/issues/34837.
I guess it's worth clarifying it a little bit :)

### CODE REVIEW

File: components/http_client.rst [hunk 1]
```diff
@@ -417,6 +418,18 @@ When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
...
     $content = $response->getContent(false);
 
+
+Thus, there are two types of exceptions:
+
+* Exceptions implementing the ``HttpExceptionInterface`` are thrown when your code
+  does not handle the status codes in the 300-599 range.
+  When streaming responses, these exceptions do not interrupt the stream as you might
+  still want to read the content of these responses.
+  That being said, be aware that they are thrown on every single chunk.
```
Comment (nicolas-grekas) [inline components/http_client.rst]: that's not accurate: chunks don't throw anything about status codes
but if you repeatidily call the same method (e.g. `$response->getHeaders()`) again and again, it's going to throw again and again for sure :)

Comment (Toflar) [reply components/http_client.rst]: Yes, that's what I meant. So wording still needs to be improved :) Proposals?

File: components/http_client.rst [hunk 2]
```diff
@@ -404,7 +404,8 @@ Handling Exceptions
 
 When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
 4xx or 5xx) your code is expected to handle it. If you don't do that, the
-``getHeaders()`` and ``getContent()`` methods throw an appropriate exception::
+``getHeaders()`` and ``getContent()`` methods throw an appropriate exception. All of
+which implement the ``HttpExceptionInterface``::
```
Comment (HeahDude) [inline components/http_client.rst]:
```suggestion
which implement the :class:`Symfony\\Contracts\\HttpClient\\Exception\\HttpExceptionInterface`::
```

**Review by HeahDude — CHANGES_REQUESTED** [review]
Thanks @Toflar for this PR. Could you please have a look at # #13136 which has been merged recently and take it into account. Also please rebase onto 4.4, since 4.3 is not maintained anymore. Thanks!

**Comment by Toflar** [comment]
Based onto 4.4 and addressed comments. I guess I don't need to rebase and force push as you're going to squash merge anyway? :)

File: components/http_client.rst [hunk 3]
```diff
@@ -477,6 +478,14 @@ errors, exceptions will notify you when needed. On the other hand, if you write
 the error-handling code, you will opt-out from these fallback mechanisms as the
 destructor won't have anything remaining to do.
 
+There are two types of exceptions:
```
Comment (nicolas-grekas) [inline components/http_client.rst]: there is also `DecodingExceptionInterface` :)

**Review by nicolas-grekas — COMMENTED** [review]
Please rebase - squashing PRs with merge commits takes more time for maintainers.

File: components/http_client.rst [hunk 4]
```diff
@@ -436,7 +436,8 @@ Handling Exceptions
 
 When the HTTP status code of the response is in the 300-599 range (i.e. 3xx,
 4xx or 5xx) your code is expected to handle it. If you don't do that, the
-``getHeaders()`` and ``getContent()`` methods throw an appropriate exception::
+``getHeaders()`` and ``getContent()`` methods throw an appropriate exception All of
```
Comment (nicolas-grekas) [inline components/http_client.rst]: missing `,` + lowercase `a`?

File: components/http_client.rst (hunk 3 above)
Comment (Toflar) [reply components/http_client.rst]: Added.

File: components/http_client.rst (hunk 4 above)
Comment (Toflar) [reply components/http_client.rst]: Fixed.

**Comment by Toflar** [comment]
Rebased :)

File: components/http_client.rst [hunk 5]
```diff
@@ -476,6 +477,29 @@ component. No errors will be unnoticed: if you don't write the code to handle
 errors, exceptions will notify you when needed. On the other hand, if you write
 the error-handling code, you will opt-out from these fallback mechanisms as the
 destructor won't have anything remaining to do.
+=======
```
Comment (nicolas-grekas) [inline components/http_client.rst]: should be removed I suppose?

File: components/http_client.rst [hunk 6]
```diff
@@ -476,6 +477,29 @@ component. No errors will be unnoticed: if you don't write the code to handle
...
+Thus, there are two types of exceptions:
+
+* Exceptions implementing the ``HttpExceptionInterface`` are thrown when your code
+  does not handle the status codes in the 300-599 range.
+  When streaming responses, these exceptions do not interrupt the stream as you might
+  still want to read the content of these responses.
+  That being said, be aware that they are thrown on every single chunk.
+
+* ``TransportExceptionInterface`` exceptions signal a lower level issue.
+  Thus, when streaming responses, the exception will only be thrown once.
```
Comment (nicolas-grekas) [inline components/http_client.rst]: all the above block should be removed, isn't it?

**Comment by Toflar** [comment]
Sorry, messed up with the rebase. Should be fixed now 🤦‍♂

**Review by nicolas-grekas — APPROVED** [review]

**Comment by javiereguiluz** [comment]
Thanks Yanick.

**Pull request merged**